### ISSUE TITLE
Interesting bundling alternative

### ISSUE THREAD

**Issue by benmccann** [roxiness/routify#222]
I wanted to share something that I've been working on that I thought you might find interesting.

I was just checking out https://routify.dev/guide/bundling and there's actually a third option I came up with, which is to split the components into chunks, but then add a `preload` header for every chunk. This gets rid of the request waterfall, but also has another benefit, which is that when you navigate to a new page it only needs to fetch the components that are new to that page because the others will have been cached.

I wrote [a small library](https://github.com/benmccann/rollup-dependency-tree) that uses the output from a Rollup compilation to figure out what files need to be preloaded for a given page. I have a PR pending over at Sapper that implements it in that project

**Comment by jakobrosenberg** [comment]
There's a preload option in Routify, but I'm not sure if it's actually documented.
``<!-- routify:options preload=true -->``
That preloads the component with the `main.js` bundle.

Are you suggesting that if you have a component chain of `layout->layout->layout->page`, that all uncached components are imported in parallel? If so, that's a very easy addition as `$page.preload()` will import it's own component as well as all its layouts.

**Comment by kevmodrome** [comment]
I'm using https://github.com/vikerman/rollup-plugin-hoist-import-deps for this on Svelte School, works great out of the box.

**Comment by jakobrosenberg** [comment]
This isn't my forte. Aren't dynamic imports imported in parallel if they're requested in parallel?

**Comment by kevmodrome** [comment]
Yes, I think so, but it's impossible to know of any static deps that the dynamic ones import. Maybe I'm misunderstanding what the purpose is of the library.

**Comment by benmccann** [comment]
Yeah, so what I'm doing is downloading the whole dependency chain for a page. The layouts, page, and components within a page.

With https://github.com/vikerman/rollup-plugin-hoist-import-deps you still have a small request waterfall. It's been awhile since I tested that plugin, but I think that the main script is loaded first and then the dynamic import and its dependencies are loaded afterwards. With my solution all are preloaded in parallel with the initial page load

**Comment by rixo** [comment]
> add a preload header for every chunk

You mean [this](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content)?

> This isn't my forte. Aren't dynamic imports imported in parallel if they're requested in parallel?

They're downloaded in parallel but browsers have a limit on the number of requests they'll send to a single domain simultaneously (ballpark 6 if I'm not mistaken). CDNs leverages this by serving from a different domain.

@benmccann So, your proposed solution is to add `<link rel="preload" href="foo.js" as="script">` for every transitive deps, static or dynamic, or an "entrypoint" chunk? How would you add that? When generating the initial HTML file that will be loaded, or rather something like a bit of client-side code that adds them all dynamically (pulling them from a build-time dictionary/map?)?

**Comment by benmccann** [comment]
> You mean this?

Basically. But added as an [HTTP Link header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link) instead of to the document's `head` element

> They're downloaded in parallel but browsers have a limit on the number of requests they'll send to a single domain simultaneously (ballpark 6 if I'm not mistaken). CDNs leverages this by serving from a different domain.

Good callout, which I hadn't considered. However, it turns out it's actually not an issue with HTTP/2. I see it loading about 9-10 files in parallel on my test page. From https://stackoverflow.com/a/36847527/64421:
With HTTP/2, browsers open only 1 connection per domain. However, thanks to the multiplexing feature of the HTTP/2 protocol, the number of concurrent requests per domain is not limited to 6-8, but it is virtually unlimited.

> So, your proposed solution is to add <link rel="preload" href="foo.js" as="script"> for every transitive deps, static or dynamic, or an "entrypoint" chunk?

Yep

> When generating the initial HTML file that will be loaded, or rather something like a bit of client-side code that adds them all dynamically (pulling them from a build-time dictionary/map?)?

The way I did it in Sapper is to have the server add the headers when loading a route from a build-time dictionary/map. I guess with Routify you couldn't take the same exact approach since there's no server component. Perhaps you'd be able to do it client-side still, but I haven't investigated that

**Comment by jakobrosenberg** [comment]
One approach I've considered is tracking the `$url` helper. The helper has exact knowledge of all outgoing links from any given page, granted that people use the `$url` helper.

I don't know if that relates to the current discussion, cause it's a bit over my head.

**Comment by Wolfr** [comment]
@benmccann I am cleaning up issues, thanks for your interesting contribution. If everyone doesn't mind, I'll close this issue in shortly.

**Comment by cortopy** [comment]
@Wolfr please don't close this issue. 

Loading of JS scripts affects so many performance metrics and Lighthouse scores that this is a crucial topic for any frontend framework. Sapper has already merged @benmccann's proposal. I personally like it very much. Even if it's not accepted, it would be nice to have an approach that may be unique to routify but is thought through too

**Comment by jakobrosenberg** [comment]
I can't tell what this adds that isn't already there. When you go to an uncached page, all required components are requested in parallel. On subsequent visits, they're served from cache. If pages need to be preloaded, it can be done at runtime with `$page.preload()` or at build with `<!-- routify:options preload=true -->`.

What am I missing here? 😳

**Comment by benmccann** [comment]
I don't quite understand what functionality is in Routify today since I don't see a whole lot in the docs, but from what I see the term "preload" is being used in a different manner.

If you go to https://realworld.svelte.dev/ then Sapper adds a header to the response:

```
link: </client/client.0422cf2a.js>;rel="modulepreload";as="script", </client/assets/client-39ff2e19.css>;rel="preload";as="style", </client/index.34c8e470.js>;rel="modulepreload";as="script", </client/api.5c9dd840.js>;rel="modulepreload";as="script", </client/index.37804c28.js>;rel="modulepreload";as="script", </client/Home.26cab2fc.js>;rel="modulepreload";as="script"
```

This causes the JS and CSS for that page to be requested immediately to avoid request waterfalls

No need to keep this open though if it's something you already have or doesn't fit with the project

**Comment by cortopy** [comment]
just to give a bit of context about metrics, I run a quick lighthouse audit on [routify.dev](https://routify.dev/). It scores rather pourly on Performance (38 out of 100). I think the biggest issue is not having critical css, but that's another issue.

In this screenshot I'm pointing at the resources which could be loaded in a more performant manner using 'preload'

![image](https://user-images.githubusercontent.com/8725825/93394762-9f3db800-f86c-11ea-8ddd-15a6482bdd8b.png)

This is the [link](https://web.dev/uses-rel-preload/?utm_source=lighthouse&utm_medium=devtools) about how lighthouse comes to this conclussion

**Comment by jakobrosenberg** [comment]
I think the poor performance is caused by lack of CDN. Couldn't get it working with Cloudflare. If I run Chrome's builtin Lighthouse, I get all 100%.

Workbox could probably be preloaded, but `/build/layout...js` and `/build/index...js` are mission critical, unless you're hydrating a static page.

For static pages it might make sense, to load the actual app, only after the network is idle.

**Comment by ghostdevv** [comment]
Can this be closed? @jakobrosenberg
